analyzers/src/SonarAnalyzer.CSharp.Styling/Rules/SeparateDeclarations.cs

```diff
@@ -0,0 +1,107 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2024 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules.CSharp.Styling;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class SeparateDeclarations : StylingAnalyzer
+{
+    public SeparateDeclarations() : base("T0016", "Add an empty line before this declaration.") { }
+
+    protected override void Initialize(SonarAnalysisContext context)
+    {
+        context.RegisterNodeAction(
+            ValidateSeparatedMember,
+            SyntaxKind.ClassDeclaration,
+            SyntaxKind.ConstructorDeclaration,
+            SyntaxKind.ConversionOperatorDeclaration,
+            SyntaxKind.DestructorDeclaration,
+            SyntaxKind.EnumDeclaration,
+            SyntaxKind.MethodDeclaration,
+            SyntaxKind.NamespaceDeclaration,
+            SyntaxKind.OperatorDeclaration,
+            SyntaxKind.RecordDeclaration,
+            SyntaxKind.RecordStructDeclaration,
+            SyntaxKind.StructDeclaration);
+        context.RegisterNodeAction(
+            ValidatePossibleSingleLineMember,
+            SyntaxKind.EventDeclaration,
+            SyntaxKind.EventFieldDeclaration,
+            SyntaxKind.DelegateDeclaration,
+            SyntaxKind.FieldDeclaration,
+            SyntaxKind.IndexerDeclaration,
+            SyntaxKind.PropertyDeclaration);
+    }
+
+    private void ValidatePossibleSingleLineMember(SonarSyntaxNodeReportingContext context)
+    {
+        var firstToken = context.Node.GetFirstToken();
+        if (firstToken.Line() != context.Node.GetLastToken().Line()
+            || firstToken.GetPreviousToken().IsKind(SyntaxKind.CloseBraceToken)
+            || PreviousDeclarationKind(context.Node) != context.Node.Kind())
+        {
+            ValidateSeparatedMember(context);
+        }
+
+        SyntaxKind PreviousDeclarationKind(SyntaxNode node) =>
+            context.Node.Parent.ChildNodes().TakeWhile(x => x != node).LastOrDefault() is { } preceding
+                ? preceding.Kind()
+                : SyntaxKind.None;
```

Comment:
```suggestion
        static SyntaxKind PreviousDeclarationKind(SyntaxNode node) =>
            node.Parent.ChildNodes().TakeWhile(x => x != node).LastOrDefault() is { } preceding
                ? preceding.Kind()
                : SyntaxKind.None;
```

Reply:
The other way around - it was a leftover from other refactoring (before I declared the abstract one a FN). 

There's no reason to make it static just to pass information that is already in scope.